File: last/toby/parsers/Parser.java
/*
 * TOBY -- A LOGO-like interpreted language, written in 100% pure Java.
 * Copyright (C) 1999  Ryan C. Gordon.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

package last.toby.parsers;
import last.toby.util.TobyLanguage;
import last.toby.util.BugSquasher;
import last.toby.interpreter.GlobalLogicContext;
import last.toby.exceptions.ParseException;
import java.io.Reader;

    /**
     *  This gives us a common interface to controlling various parsers.
     *
     *   @author Ryan C. Gordon. (icculus@lokigames.com)
     */
public abstract class Parser
{
    protected long totalParseTime = -1;
    protected long totalLinkTime = -1;

        /**
         * If debugging is enabled, then println debug info to stdout.
         *
         *   @param funcName name of method where debugging info lies.
         *   @param msg Message to print.
         */
    protected void _D(String funcName, String msg)
    {
        BugSquasher.println(funcName, msg);
    } // _D


        /**
         * Works like a C++ assertion.
         *
         *   assert(2 > 4) will throw a ParseException, since it's false.
         *   assert(2 < 4) will not.
         *
         *   @param expression true or false expression.
         *   @throw ParseException if (expression) is true.
         */
    public static void assert(boolean expression) throws ParseException
    {
        BugSquasher.assert(expression);
    } // assert

        /**
         * A simple bubble sort to alphabetize an array of strings.
         *
         *   @param array Array to sort.
         *  @return same reference to (array) that was passed as a
         *          param, sorted.
         */
    public static String[] alphabetizeArrayElements(String[] array)
    {
        int i;
        int j;
        int max;
        String tmp;
        boolean sorted;

        if (array == null)      // safety check.
            return(null);

        max = array.length - 1;

        do
        {
            sorted = true;
            for (i = 0; i < max; i++)
            {
                if (array[i].compareTo(array[i + 1]) > 0)
                {
                    sorted = false;
                    tmp = array[i];
                    array[i] = array[i + 1];
                    array[i + 1] = tmp;
                } // if
            } // for
        } while (sorted == false);

        return(array);
    } // alphabetizeArrayElements


        /**
         * Find out if (forThis) is in (table). The strings contained in table
         *  must be indexed in alphabetic order. This makes the searches faster.
         *
         *     params : see above.
         *    returns : index in table, if there. (-1) if (forThis) isn't there.
         */
    public static int searchAlphabeticArray(String[] table, String forThis)
    {
        int i;
        int rc;
        int max = table.length;

        if (forThis != null)   // bullet proofing.
        {
            for (i = 0; i < max; i++)
            {
                rc = forThis.compareTo(table[i]);
                if (rc == 0)
                    return(i);
                else if (rc < 0)  // alphabetically past search string?
                    return(-1);
            } // for
        } // if

        return(-1);   // not found. Bail.
    } // searchAlphabeticArray


        /**
         * Find out if (forThis) is in (table).
         *
         *     params : see above.
         *    returns : index in table, if there. (-1) if (forThis) isn't there.
         */
    public static int searchArray(String[] table, String forThis)
    {
        int i;
        int max = table.length;

        if (forThis != null)
        {
            for (i = 0; i < max; i++)
            {
                if (table[i] != null)
                {
                    if (table[i].compareTo(forThis) == 0)
                        return(i);
                } // if
            } // for
        } // if
        return(-1);
    } // searchArray


        /**
         * Find out if (forThis) is in (table). (forThis) must not be just
         *  an identical string, but an identical reference.
         *
         *     params : see above.
         *    returns : index in table, if there. (-1) if (forThis) isn't there.
         */
    public static int searchArrayForReference(String[] table, String forThis)
    {
        int i;

        for (i = 0; i < table.length; i++)
        {
            if (table[i] == forThis)
                return(i);
        } // for

        return(-1);
    } // searchArrayForReference




        /**
         * Call this to parse a collection of code into a LogicContext tree.
         *  (code) may be an entire program or just a single procedure or
         *  whatever else.
         *
         *   @param code The code to parse. This is usually a StringReader,
         *               but may come directly from disk or network.
         *   @param screen An object representing the screen. This is
         *                 language/platform, etc dependent, but in the
         *                 usual/intended case, it's an implementor of
         *                 last.toby.turtlespace.TurtleSpace.
         *  @return A new ModuleLogicContext containing procedures, global
         *          vars, and other attributes resulting from the parsing of
         *          (code).
         *   @throw ParseException on syntax errors, etc...
         */
    public final GlobalLogicContext parseCode(Reader code, Object screen)
                                                throws ParseException
    {
//        _D("parseCode", "Starting parsing phase...");
        long parseStartTime = System.currentTimeMillis();
        GlobalLogicContext retVal = parseCodeImpl(code);
        this.totalParseTime = System.currentTimeMillis() - parseStartTime;

        if (retVal != null)
        {
            retVal.setScreen(screen);

//            _D("parseCode", "Starting linking phase...");
            long linkStartTime = System.currentTimeMillis();
            retVal.link(retVal);
            this.totalLinkTime = System.currentTimeMillis() - linkStartTime;
        } // if

//        _D("parseCode", "Total parse time: (" + this.totalParseTime + ") ms.");
//        _D("parseCode", "Total link  time: (" + this.totalLinkTime  + ") ms.");

        return(retVal);
    } // parseCode


        /**
         * Get the total time that was spent during the last parsing,
         *  in milliseconds.
         *
         *   @return total parse time, in milliseconds.
         */
    public final long getTotalParseTime()
    {
        return(this.totalParseTime);
    } // getTotalParseTime


        /**
         * Get the total time that was spent during the last linking,
         *  in milliseconds.
         *
         *   @return total link time, in milliseconds.
         */
    public final long getTotalLinkTime()
    {
        return(this.totalLinkTime);
    } // getTotalLinkTime

        /**
         * This is the entry point for a given parser to do its work.
         *
         *    @param The entire module's source code as one big string.
         *            Newlines included.
         *   @return a GlobalLogicContext containing the parsed code's
         *           logic structure.
         */
    public abstract GlobalLogicContext parseCodeImpl(Reader code)
                                                     throws ParseException;
} // Parser

// end of Parser.java ...

File: src/parsers/Parser.cpp
/*
 * TOBY -- An abstract interpreter engine and system for learning.
 * Copyright (C) 1999  Ryan C. Gordon.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include "parsers/Parser.h"

Parser::Parser(void)
    : totalParseTime(0xdeadbeef),
      totalLinkTime(0xdeadbeef)
{
} // Constructor


Parser::~Parser(void)
{
} // Destructor


    // !!! Change this bubble sort with something faster?
TobyCollection *Parser::alphabetizeArrayElements(TobyCollection *array)
{
    int i;
    int max;
    bool sorted;

    if (array == NULL)      // safety check.
        return(NULL);

    max = array->getSize() - 1;

    if (max <= 0)
        return(array);

    do
    {
        sorted = true;
        for (i = 0; i < max; i++)
        {
            char *str1 = (char *) array->elementAt(i);
            char *str2 = (char *) array->elementAt(i + 1);
            if (strcmp(str1, str2) > 0)
            {
                sorted = false;
                array->swapElements(i, i + 1);
            } // if
        } // for
    } while (sorted == false);

    return(array);
} // Parser::alphabetizeArrayElements


// !!! this is a slow algorithm.
int Parser::searchAlphabeticArray(TobyCollection *table, const char *forThis)
{
    if (forThis != NULL)   // bullet proofing.
    {
        int max = table->getSize();
        for (int i = 0; i < max; i++)
        {
            int rc = strcmp(forThis, (char *) (table->elementAt(i)));
            if (rc == 0)
                return(i);
            else if (rc < 0)  // alphabetically past search string?
                return(-1);
        } // for
    } // if

    return(-1);   // not found. Bail.
} // Parser::searchAlphabeticArray


int Parser::searchArray(TobyCollection *table, const char *forThis)
{
    if (forThis != NULL)
    {
        int max = table->getSize();
        for (int i = 0; i < max; i++)
        {
            if (strcmp((char *) table->elementAt(i), forThis) == 0)
                return(i);
        } // for
    } // if

    return(-1);
} // Parser::searchArray


int Parser::searchArrayForReference(TobyCollection *table, const char *str)
{
    int max = table->getSize();

    for (int i = 0; i < max; i++)
    {
        if (table->elementAt(i) == (void *) str)
            return(i);
    } // for

    return(-1);
} // Parser::searchArrayForReference


GlobalLogicContext *Parser::parseCode(TobyReader *code, TobyObject *screen)
                                                throw (ParseException *)
{
    _D(("Parser::parseCode(): Starting parsing phase...\n"));
    TobyClock *clock = TobyClock::getSingleton();
    toby_uint32 parseStartTime = clock->getTicks();
    GlobalLogicContext *retVal = parseCodeImpl(code);
    totalParseTime = clock->getTicks() - parseStartTime;

    if (retVal != NULL)
    {
        // !!! retVal->setScreen(screen);

        _D(("Parser::parseCode(): Starting linking phase...\n"));
        toby_uint32 linkStartTime = clock->getTicks();
        // !!! retVal->link(retVal);
        totalLinkTime = clock->getTicks() - linkStartTime;
    } // if

    _D(("Parser::parseCode(): Total parse time: (%lu) ms.\n", totalParseTime));
    _D(("Parser::parseCode(): Total link  time: (%lu) ms.\n", totalLinkTime));

    return(retVal);
} // Parser::parseCode


toby_uint32 Parser::getTotalParseTime(void)
{
    return(totalParseTime);
} // Parser::getTotalParseTime


toby_uint32 Parser::getTotalLinkTime(void)
{
    return(totalLinkTime);
} // Parser::getTotalLinkTime


// end of Parser.cpp ...

